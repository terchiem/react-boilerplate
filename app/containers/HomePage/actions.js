import {
  GET_POSTS,
  GET_POSTS_SUCCESS,
  GET_POSTS_FAIL,
  ADD_POST_TO_LIST,
} from './constants';

export function getPosts() {
  return {
    type: GET_POSTS,
  };
}

export function getPostsSuccess(posts) {
  return {
    type: GET_POSTS_SUCCESS,
    posts,
  };
}

export function getPostsFail(error) {
  return {
    type: GET_POSTS_FAIL,
    error,
  };
}

export function addPostToList(post) {
  return {
    type: ADD_POST_TO_LIST,
    post,
  };
}
