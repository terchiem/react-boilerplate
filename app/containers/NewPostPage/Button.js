import styled from 'styled-components';

const Button = styled.button`
  padding: 0.25em 2em;
  margin: 1em;
  text-decoration: none;
  border-radius: 4px;
  user-select: none;
  cursor: pointer;
  font-weight: bold;
  font-size: 16px;
  border: 2px solid #333;
  background: #fff;
  color: #333;

  &:hover {
    background: #333;
    color: #fff;
  }

  &:active {
    background: #888;
    border: 2px solid #888;
  }
`;

export default Button;
