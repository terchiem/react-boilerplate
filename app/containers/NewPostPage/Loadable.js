/**
 * Asynchronously loads the component for NewPostPage
 */

import loadable from 'utils/loadable';

export default loadable(() => import('./index'));
