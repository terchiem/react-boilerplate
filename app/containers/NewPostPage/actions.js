import { CHANGE_POST_INPUT, RESET_POST_INPUT } from './constants';

export function changePostInput(value) {
  return {
    type: CHANGE_POST_INPUT,
    value,
  };
}

export function resetPostInput() {
  return {
    type: RESET_POST_INPUT,
  };
}
