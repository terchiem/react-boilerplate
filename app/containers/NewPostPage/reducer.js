import produce from 'immer';
import { CHANGE_POST_INPUT, RESET_POST_INPUT } from './constants';

export const INITIAL_STATE = {
  inputValue: '',
};

function newPostReducer(state = INITIAL_STATE, action) {
  return produce(state, draft => {
    switch (action.type) {
      case CHANGE_POST_INPUT:
        draft.inputValue = action.value;
        break;

      case RESET_POST_INPUT:
        draft.inputValue = '';
        break;

      default:
        break;
    }
    return draft;
  });
}

export default newPostReducer;
