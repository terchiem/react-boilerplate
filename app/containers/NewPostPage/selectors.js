/**
 * NewPostPage selectors
 */

import { createSelector } from 'reselect';
import { INITIAL_STATE } from './reducer';

export const selectNewPost = state => state.newPost || INITIAL_STATE;

export const makeSelectInputValue = () =>
  createSelector(
    selectNewPost,
    state => state.inputValue,
  );
