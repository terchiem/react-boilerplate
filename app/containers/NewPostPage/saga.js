/**
 * Saga to add a new post to the server through API call
 */

import request from 'utils/request';
import { push } from 'connected-react-router';
import { call, put, takeLatest, select } from 'redux-saga/effects';
import { ADD_POST } from 'containers/HomePage/constants';
import { addPostSuccess, addPostFail } from 'containers/HomePage/actions';
import { resetPostInput } from './actions';
import { makeSelectInputValue } from './selectors';

export function* addNewPost() {
  const requestURL = `http://localhost:3000/api`;

  try {
    const post = yield select(makeSelectInputValue());
    const response = yield call(request, requestURL, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ post }),
    });
    yield put(addPostSuccess(response.post));
    yield put(resetPostInput());
    yield put(push('/'));
  } catch (err) {
    yield put(addPostFail(err));
  }
}

export default function* addPostSaga() {
  yield takeLatest(ADD_POST, addNewPost);
}
