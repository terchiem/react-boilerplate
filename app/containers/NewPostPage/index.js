import React from 'react';
import PropTypes from 'prop-types';
import history from 'utils/history';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { createStructuredSelector } from 'reselect';
import injectReducer from 'utils/injectReducer';
import injectSaga from 'utils/injectSaga';
import { addPost } from 'containers/HomePage/actions';
import { makeSelectInputValue } from './selectors';
import { changePostInput, resetPostInput } from './actions';
import reducer from './reducer';
import saga from './saga';

import Button from './Button';

/*
 * NewPostPage
 *
 * Displays a form for a user to create a new post, at the '/new' route
 *
 * Redux state:
 *  newPost: {
 *    inputValue -> current value of the text input field for newPost
 *  }
 *
 */

function NewPostPage({ inputValue, handleChange, handleSubmit }) {
  return (
    <section>
      <h2>New Post</h2>
      <form onSubmit={handleSubmit}>
        <input
          id="newPost"
          value={inputValue}
          onChange={handleChange}
          aria-label="newPost"
          placeholder="Say something..."
        />
        <Button type="submit">Submit</Button>
      </form>
    </section>
  );
}

NewPostPage.propTypes = {
  inputValue: PropTypes.string,
  handleChange: PropTypes.func,
  handleSubmit: PropTypes.func,
};

const mapStateToProps = createStructuredSelector({
  inputValue: makeSelectInputValue(),
});

const mapDispatchToProps = dispatch => ({
  handleChange: evt => dispatch(changePostInput(evt.target.value)),
  handleSubmit: evt => {
    evt.preventDefault();
    dispatch(addPost(evt.target.value));
    dispatch(resetPostInput());
    history.push('/');
  },
});

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);
const withReducer = injectReducer({ key: 'newPost', reducer });
const withSaga = injectSaga({ key: 'newPost', saga });

export default compose(
  withConnect,
  withReducer,
  withSaga,
)(NewPostPage);
