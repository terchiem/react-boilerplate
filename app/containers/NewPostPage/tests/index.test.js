import React from 'react';
import { Provider } from 'react-redux';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
import { browserHistory } from 'react-router-dom';
import { addPost } from 'containers/HomePage/actions';
import NewPostPage, { mapDispatchToProps } from '../index';
import { changePostInput } from '../actions';
import configureStore from '../../../configureStore';

describe('<NewPostPage />', () => {
  let store;

  beforeAll(() => {
    store = configureStore({}, browserHistory);
  });

  it('should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <NewPostPage />
        </IntlProvider>
      </Provider>,
    );
    expect(firstChild).toMatchSnapshot();
  });

  describe('mapDispatchToProps', () => {
    describe('handleChange', () => {
      it('should be injected', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        expect(result.handleChange).toBeDefined();
      });

      it('should dispatch changeUsername when called', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        const value = 'Test';
        result.handleChange({ target: { value } });
        expect(dispatch).toHaveBeenCalledWith(changePostInput(value));
      });
    });

    describe('handleSubmit', () => {
      it('should be injected', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        expect(result.handleSubmit).toBeDefined();
      });

      it('should dispatch addPost when called', () => {
        const dispatch = jest.fn();
        const preventDefault = jest.fn();
        const result = mapDispatchToProps(dispatch);
        const evt = { preventDefault };
        result.handleSubmit(evt);
        expect(dispatch).toHaveBeenCalledWith(addPost());
      });
    });
  });
});
