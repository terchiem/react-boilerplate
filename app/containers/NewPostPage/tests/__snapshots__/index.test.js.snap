// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewPostPage /> should render and match the snapshot 1`] = `
<section>
  <h2>
    New Post
  </h2>
  <form>
    <input
      aria-label="newPost"
      id="newPost"
      placeholder="Say something..."
      required=""
      value=""
    />
    <button
      class="Button-l2b7vc-0 iiHOpd"
      type="submit"
    >
      Submit
    </button>
  </form>
</section>
`;
